File: ClaudeBattery/ClaudeBattery/Views/DialForecast.swift
import SwiftUI

extension UsagePopoverView {
    // MARK: - Ring colour (#46, KD5, KD13)

    /// Outer ring colour: red below `UsageData.lowRemainingThreshold` whatever the pace says (the
    /// menu-bar icons go red at the same number, so icon and popover agree about "nearly empty"),
    /// else the pace colour so ring and word agree, else the level colour when there is no pace to
    /// follow (`.weeklyLimited`, `.unknown`). The Session card passes the DISPLAY value here, the
    /// same number the renderers test (KTD2).
    static func ringColor(remaining: Double, pace: PaceStatus) -> Color {
        if remaining < UsageData.lowRemainingThreshold { return .red }
        switch pace {
        case .onTrack, .caution, .danger: return paceColor(pace)
        case .weeklyLimited, .unknown:    return batteryColor(remainingPercent: remaining)
        }
    }

    /// Colour of the pace word under the dial: the ring's red floor, otherwise `paceColor`, so the
    /// word can never disagree with the ring and "Limited by weekly" keeps its muted colour above
    /// the floor (KD13). Takes the same value the ring does.
    static func paceCaptionColor(remaining: Double, pace: PaceStatus) -> Color {
        if remaining < UsageData.lowRemainingThreshold { return .red }
        return paceColor(pace)
    }

    /// The inner time ring is a fixed neutral grey (KD8): lighter than `trackColor` so it reads as
    /// a ring, and not cyan or blue, which already means "spend is fine" on the credits row.
    static let neutralInnerRingColor = Color(white: 0.5)

    // MARK: - Run-out estimate (#31, KD6, KD7, KTD1)

    /// Below this fraction of the window elapsed no run-out is projected: a burst of use at the
    /// start of a fresh window divides by a tiny elapsed time and projects a run-out minutes away.
    static let minElapsedFraction: Double = 0.10

    /// Seconds until usage runs out at the current burn rate, from the current snapshot only, or
    /// nil when there is nothing sound to project. `pace` is the result of `paceStatus` on the SAME
    /// inputs: the projection reads it instead of re-deriving the delta, so the word and the
    /// estimate can never drift apart, and it returns nil unless the pace is `.caution` or
    /// `.danger`. That gate already implies at least 10% used and at most 90% left, so the only
    /// guards kept are the ones it does not imply: depleted (nothing left to project), under
    /// `minElapsedFraction` of the window elapsed, and a reset beyond one window (clock skew makes
    /// the elapsed time negative). Window start is `resetsAt - window`, as in `timeRemainingPercent`.
    /// The formula `(window - r) * remaining / (100 - remaining)` is below `r` exactly when the
    /// pace delta is positive, so the estimate always lands before the reset.
    static func runOutSeconds(remainingPercent: Double, resetsAt: Date?, window: TimeInterval,
                              pace: PaceStatus, now: Date = Date()) -> TimeInterval? {
        guard pace == .caution || pace == .danger else { return nil }
        guard let resetsAt,
              let r = CountdownFormat.remainingSeconds(until: resetsAt, now: now) else { return nil }
        if remainingPercent <= 0 { return nil }
        if r > window { return nil }
        let elapsedFraction = 1 - r / window
        if elapsedFraction < minElapsedFraction { return nil }
        return (window - r) * remainingPercent / (100 - remainingPercent)
    }

    /// Rounds a run-out to the nearest 5 minutes on the session window and the nearest hour on
    /// the weekly window, never below one quantum, so poll-to-poll jitter in the integer
    /// percentages does not move the shown value.
    static func quantiseRunOut(_ seconds: TimeInterval, window: TimeInterval) -> TimeInterval {
        let quantum: TimeInterval = window == weeklyWindow ? 3600 : 300
        return max(1, (seconds / quantum).rounded()) * quantum
    }

    /// Copy for the run-out line under the pace word, relative so it cannot be mistaken for a
    /// reset time (KD6). Prints the seconds it is given; the caller quantises first. nil hides it.
    static func runOutLine(seconds: TimeInterval?) -> String? {
        guard let seconds else { return nil }
        return "Out in ~" + CountdownFormat.minuteResolution(seconds: seconds)
    }

    // MARK: - Dial lines (#44, KTD3)

    /// The three text lines under a dial, in display order: the pace word (nil hides it), the
    /// run-out estimate (nil hides it), and the reset countdown, which is always present. The two
    /// durations behind the strings ride along for the spoken label; `runOutSeconds` is already
    /// quantised, so the voice and the line agree.
    struct DialLines: Equatable {
        let caption: String?
        let runOut: String?
        let countdown: String
        let resetSeconds: TimeInterval?
        let runOutSeconds: TimeInterval?
    }

    /// The single line a dial shows when its reset time is nil or already past (AE5).
    static let resetUnavailableLine = "Reset time unavailable"

    /// The lines under one dial from the current snapshot (KTD3). Takes no display value and no
    /// weekly-limited flag: `pace` already encodes weekly-limited-versus-Danger through
    /// `sessionPace(for:)`, and `rawRemaining` is the value the run-out projects. A nil or past
    /// reset self-omits everything but "Reset time unavailable", whatever `pace` says, matching the
    /// dial's own inner-arc rule (KTD4). The run-out is quantised here, before `runOutLine`, which
    /// prints exactly what it is given, and then dropped unless it still lands strictly before the
    /// reset: `pace` is graded at poll time while these lines re-print on the minute clock, so a
    /// stale Caution can otherwise project past a reset the countdown says is nearer.
    static func dialLines(pace: PaceStatus, rawRemaining: Double, resetsAt: Date?, window: TimeInterval,
                          now: Date = Date()) -> DialLines {
        guard let resetsAt,
              let resetSeconds = CountdownFormat.remainingSeconds(until: resetsAt, now: now) else {
            return DialLines(caption: nil, runOut: nil, countdown: resetUnavailableLine,
                             resetSeconds: nil, runOutSeconds: nil)
        }
        let runOut = runOutSeconds(remainingPercent: rawRemaining, resetsAt: resetsAt, window: window,
                                   pace: pace, now: now)
            .map { quantiseRunOut($0, window: window) }
            .flatMap { $0 < resetSeconds ? $0 : nil }
        return DialLines(caption: paceCaption(pace),
                         runOut: runOutLine(seconds: runOut),
                         countdown: "Resets in " + CountdownFormat.minuteResolution(seconds: resetSeconds),
                         resetSeconds: resetSeconds,
                         runOutSeconds: runOut)
    }

    /// The two inputs the Session card's lines project from: the pace from `sessionPace(for:)`
    /// and the RAW `sessionRemaining`, picked here and nowhere else, so neither `sessionCard` nor
    /// any other caller can hand the weekly-capped display value to the run-out (KD7).
    static func sessionDialInputs(for usage: UsageData, now: Date = Date()) -> (pace: PaceStatus, rawRemaining: Double) {
        (pace: sessionPace(for: usage, now: now), rawRemaining: usage.sessionRemaining)
    }

    /// The Session card's lines, from `sessionDialInputs` (KD7).
    static func sessionDialLines(for usage: UsageData, now: Date = Date()) -> DialLines {
        let inputs = sessionDialInputs(for: usage, now: now)
        return dialLines(pace: inputs.pace,
                         rawRemaining: inputs.rawRemaining,
                         resetsAt: usage.sessionResetDate,
                         window: sessionWindow,
                         now: now)
    }

    /// A duration in words for the spoken label ("2 hours 30 minutes", "3 days", "less than a
    /// minute"), truncated to whole units exactly like `CountdownFormat.minuteResolution`, so the
    /// voice never says a number the line does not print. Zero components are dropped.
    static func spokenDuration(seconds: TimeInterval) -> String {
        let c = CountdownFormat.components(seconds: seconds)
        let units: [(Int, String)] = [(c.days, "day"), (c.hours, "hour"), (c.minutes, "minute")]
        let parts = units.filter { $0.0 > 0 }.map { "\($0.0) \($0.1)\($0.0 == 1 ? "" : "s")" }
        return parts.isEmpty ? "less than a minute" : parts.joined(separator: " ")
    }

    /// One spoken label per dial combining usage, time-remaining, the pace status (R9) and, when
    /// `lines` is given, the run-out and the reset countdown from under the dial, so each dial is
    /// spoken once and the three text lines can hide from the accessibility tree (KTD3). The pace
    /// is spelled out with its meaning ("over pace") rather than reusing the terse visual caption,
    /// and `.unknown` has nothing to add.
    ///
    /// `.toNearestOrEven` because these are the same two Doubles the gauge draws with "%.0f", and
    /// that is what "%.0f" does at an exact half: plain `.rounded()` would speak 17 against a dial
    /// printing 16. Every menu bar renderer already rounds this way for the same reason. #43 was
    /// this class of bug between two surfaces, and a screen reader is a third one.
    static func gaugeAccessibilityLabel(name: String, usage: Double, timeRemaining: Double?, pace: PaceStatus,
                                        lines: DialLines? = nil) -> String {
        var parts = ["\(name) usage \(Int(usage.rounded(.toNearestOrEven))) percent"]
        if let timeRemaining {
            parts.append("time remaining \(Int(timeRemaining.rounded(.toNearestOrEven))) percent")
        }
        switch pace {
        case .onTrack:       parts.append("on track")
        case .caution:       parts.append("caution, over pace")
        case .danger:        parts.append("danger, over pace")
        case .weeklyLimited: parts.append("limited by weekly")
        case .unknown:       break
        }
        if let lines {
            if let runOut = lines.runOutSeconds {
                parts.append("projected to run out in about \(spokenDuration(seconds: runOut))")
            }
            if let reset = lines.resetSeconds {
                parts.append("resets in \(spokenDuration(seconds: reset))")
            } else {
                parts.append(resetUnavailableLine.lowercased())
            }
        }
        return parts.joined(separator: ", ")
    }
}

// MARK: - Countdown Formatting

/// Shared, testable countdown helpers (U1, KTD1). Both the popover dial countdown and the
/// menu-bar compact countdown route through `remainingSeconds` so they inherit the same
/// issue-#23 trap protection: a past, non-finite, or absurdly large date yields nil rather
/// than reaching `Int(...)`, which traps fatally.
enum CountdownFormat {
    /// Seconds until `date`, or nil when there is no positive, finite, in-range countdown.
    static func remainingSeconds(until date: Date, now: Date = Date()) -> TimeInterval? {
        let remaining = date.timeIntervalSince(now)
        guard remaining > 0, remaining.isFinite, remaining < Double(Int.max) else { return nil }
        return remaining
    }

    /// Compact menu-bar countdown, never more than 3 characters (enforced for ALL inputs):
    /// `>= 10h` -> `"9h+"` (saturates; still truthful as "at least 9h" and guarantees <= 3 chars
    /// under clock skew / stale reset dates), `>= 1h` -> `"Nh+"` (e.g. "4h+"), `>= 1m` -> `"Nm"`
    /// (e.g. "32m"), `> 0 but < 1m` -> `"<1m"`. Returns nil when there is nothing to count down.
    static func compactCountdown(until date: Date, now: Date = Date()) -> String? {
        guard let remaining = remainingSeconds(until: date, now: now) else { return nil }
        let total = Int(remaining)
        let hours = total / 3600
        let minutes = (total % 3600) / 60
        if hours >= 10 { return "9h+" }
        if hours >= 1 { return "\(hours)h+" }
        if minutes >= 1 { return "\(minutes)m" }
        return "<1m"
    }

    /// Whole days, hours and minutes of a duration, truncated (`Int(seconds)`), shared by the
    /// dial countdown and the spoken label so both print the same numbers for the same instant.
    static func components(seconds: TimeInterval) -> (days: Int, hours: Int, minutes: Int) {
        let total = Int(seconds)
        return (total / 86400, (total % 86400) / 3600, (total % 3600) / 60)
    }

    /// Minute-resolution countdown for the dial lines (KD10): `>= 1d` -> `"Nd HHh"` (e.g.
    /// "3d 00h"), `>= 1h` -> `"Nh MMm"` (e.g. "2h 14m"), `>= 1m` -> `"Nm"` (e.g. "5m"), under a
    /// minute `"<1m"` so this and `compactCountdown` print the same thing for the same instant.
    /// Takes seconds so the run-out line can print a duration that has no date.
    static func minuteResolution(seconds: TimeInterval) -> String {
        let (days, hours, minutes) = components(seconds: seconds)
        if days > 0 { return String(format: "%dd %02dh", days, hours) }
        if hours > 0 { return String(format: "%dh %02dm", hours, minutes) }
        if minutes >= 1 { return "\(minutes)m" }
        return "<1m"
    }
}

File: windows/src/ClaudeBatteryWin/ViewModels/DialForecast.cs
using ClaudeBatteryWin.Icons;

namespace ClaudeBatteryWin.ViewModels;

/// <summary>
/// How fast usage is going compared with the clock (R27).
///
/// A percentage on its own does not answer the question a user actually has. Forty percent left
/// sounds bad with six hours of the window to go and is fine with twenty minutes. This grades the
/// gap between the two, which is exactly the gap between the two rings on the dial.
/// </summary>
public enum PaceStatus
{
    /// <summary>Using no faster than the clock, or slower. Includes being ahead.</summary>
    OnTrack,

    /// <summary>Running ahead of the clock, enough to be worth noticing.</summary>
    Caution,

    /// <summary>Well ahead of the clock, or nothing left at all.</summary>
    Danger,

    /// <summary>The weekly quota is the binding limit, so the session window has no pace to grade.</summary>
    WeeklyLimited,

    /// <summary>No reset time, so there is nothing to compare against. The caption is hidden.</summary>
    Unknown,
}

/// <summary>
/// The rules behind what a dial says: the pace word, the colours, the run-out estimate, and the
/// three lines underneath (R27, R28, R51, R52, R53). Ported from the Mac rule for rule.
///
/// Everything here is a pure function of one snapshot plus the current time, so the panel can
/// re-print on a one-minute clock without re-polling, and every value is testable without a window.
/// </summary>
public static class DialForecast
{
    /// <summary>Session window: five hours, which is why the session dial has five notches.</summary>
    public const double SessionWindowSeconds = 5 * 3600;

    /// <summary>Weekly window: seven days.</summary>
    public const double WeeklyWindowSeconds = 7 * 24 * 3600;

    /// <summary>At or above this many points ahead of the clock, the pace is Caution.</summary>
    public const double CautionDelta = 10;

    /// <summary>At or above this many points ahead of the clock, the pace is Danger.</summary>
    public const double DangerDelta = 25;

    /// <summary>Below this much remaining, the ring is red whatever the pace says. The tray icon
    /// goes red at the same number, so the icon and the panel agree about "nearly empty".</summary>
    public const double LowRemainingThreshold = 20;

    /// <summary>
    /// Below this fraction of the window elapsed, no run-out is projected. A burst of use in the
    /// first minutes of a fresh window divides by almost no elapsed time and projects a run-out
    /// minutes away, which is alarming and wrong.
    /// </summary>
    public const double MinElapsedFraction = 0.10;

    /// <summary>The line a dial shows instead of a countdown when it has no reset time.</summary>
    public const string ResetUnavailableLine = "Reset time unavailable";

    // --- Time and pace --------------------------------------------------------------------------

    /// <summary>
    /// How much of the window is left, as a percentage counting down to zero. The response gives
    /// only the reset time, so the window start is that minus the window length.
    /// </summary>
    public static double? TimeRemainingPercent(DateTimeOffset? resetsAt, double windowSeconds, DateTimeOffset now)
    {
        if (resetsAt is not { } reset || CountdownFormat.RemainingSeconds(reset, now) is not { } remaining)
        {
            return null;
        }

        return Math.Max(0, Math.Min(100, remaining / windowSeconds * 100));
    }

    /// <summary>
    /// The pace for one window. The delta is how much longer the inner time ring is than the outer
    /// usage ring, so the word names a gap the user can see. Nothing left is Danger whatever the
    /// clock says; being ahead of pace is always On Track.
    /// </summary>
    public static PaceStatus Pace(double remainingPercent, DateTimeOffset? resetsAt, double windowSeconds, DateTimeOffset now)
    {
        if (resetsAt is not { } reset || CountdownFormat.RemainingSeconds(reset, now) is not { } remaining)
        {
            return PaceStatus.Unknown;
        }
        if (remainingPercent <= 0)
        {
            return PaceStatus.Danger;
        }

        var timeRemaining = Math.Min(100, remaining / windowSeconds * 100);
        var delta = timeRemaining - remainingPercent;

        if (delta >= DangerDelta)
        {
            return PaceStatus.Danger;
        }
        return delta >= CautionDelta ? PaceStatus.Caution : PaceStatus.OnTrack;
    }

    /// <summary>
    /// The Session dial's pace. When the week is the binding limit the session defers to it, unless
    /// the true five-hour window is itself in Danger: that is a nearer wall and must not be hidden
    /// behind "Limited by weekly".
    ///
    /// The pace always grades the raw session value over the five-hour window, never the
    /// weekly-capped display value, which is a weekly percentage on a weekly clock and would invent
    /// a pace that means nothing.
    /// </summary>
    public static PaceStatus SessionPace(Models.UsageReading reading, DateTimeOffset now)
    {
        var raw = Pace(reading.Snapshot.SessionRemaining, reading.Snapshot.SessionResetDate, SessionWindowSeconds, now);
        if (reading.IsSessionWeeklyLimited)
        {
            return raw == PaceStatus.Danger ? raw : PaceStatus.WeeklyLimited;
        }
        return raw;
    }

    /// <summary>The word under the dial. Null hides the line.</summary>
    public static string? PaceCaption(PaceStatus status) => status switch
    {
        PaceStatus.OnTrack => "On Track",
        PaceStatus.Caution => "Caution",
        PaceStatus.Danger => "Danger",
        PaceStatus.WeeklyLimited => "Limited by weekly",
        _ => null
    };

    // --- Colours --------------------------------------------------------------------------------

    /// <summary>The plain level scale: red, orange, green by how much is left.</summary>
    public static UsageColor BatteryColor(double remainingPercent)
    {
        var clamped = Math.Max(0, Math.Min(100, remainingPercent));
        if (clamped < LowRemainingThreshold)
        {
            return UsageColor.Red;
        }
        return clamped < 45 ? UsageColor.Orange : UsageColor.Green;
    }

    /// <summary>The colour of the pace word.</summary>
    public static UsageColor PaceColor(PaceStatus status) => status switch
    {
        PaceStatus.OnTrack => UsageColor.Green,
        PaceStatus.Caution => UsageColor.Orange,
        PaceStatus.Danger => UsageColor.Red,
        _ => UsageColor.Muted
    };

    /// <summary>
    /// The outer ring's colour (R53): red below the low threshold whatever the pace says, otherwise
    /// the pace colour so the ring and the word agree, and the level colour when there is no pace to
    /// follow.
    /// </summary>
    public static UsageColor RingColor(double remaining, PaceStatus pace)
    {
        if (remaining < LowRemainingThreshold)
        {
            return UsageColor.Red;
        }

        return pace switch
        {
            PaceStatus.OnTrack or PaceStatus.Caution or PaceStatus.Danger => PaceColor(pace),
            _ => BatteryColor(remaining)
        };
    }

    /// <summary>
    /// The pace word's colour: the ring's red floor first, so the word can never disagree with the
    /// ring, and "Limited by weekly" keeps its muted colour above the floor.
    /// </summary>
    public static UsageColor PaceCaptionColor(double remaining, PaceStatus pace) =>
        remaining < LowRemainingThreshold ? UsageColor.Red : PaceColor(pace);

    // --- Run-out --------------------------------------------------------------------------------

    /// <summary>
    /// Seconds until usage runs out at the current rate, or null when there is nothing sound to
    /// project (R51). It reads the pace rather than re-deriving it, so the word and the estimate
    /// cannot drift apart, and it only projects when the pace already says usage is ahead of the
    /// clock, which is what makes the estimate land before the reset.
    /// </summary>
    public static double? RunOutSeconds(
        double remainingPercent, DateTimeOffset? resetsAt, double windowSeconds, PaceStatus pace, DateTimeOffset now)
    {
        if (pace is not (PaceStatus.Caution or PaceStatus.Danger))
        {
            return null;
        }
        if (resetsAt is not { } reset || CountdownFormat.RemainingSeconds(reset, now) is not { } remaining)
        {
            return null;
        }
        if (remainingPercent <= 0 || remaining > windowSeconds)
        {
            return null;
        }

        var elapsedFraction = 1 - remaining / windowSeconds;
        if (elapsedFraction < MinElapsedFraction)
        {
            return null;
        }

        return (windowSeconds - remaining) * remainingPercent / (100 - remainingPercent);
    }

    /// <summary>
    /// Rounds a run-out to five minutes on the session window and an hour on the weekly one, never
    /// below one of those, so the shown value does not jitter with each poll's integer percentages.
    /// </summary>
    public static double QuantiseRunOut(double seconds, double windowSeconds)
    {
        var quantum = windowSeconds == WeeklyWindowSeconds ? 3600.0 : 300.0;
        return Math.Max(1, Math.Round(seconds / quantum, MidpointRounding.ToEven)) * quantum;
    }

    /// <summary>The run-out line, relative so it cannot be mistaken for a reset time.</summary>
    public static string? RunOutLine(double? seconds) =>
        seconds is { } value ? "Out in ~" + CountdownFormat.MinuteResolution(value) : null;

    // --- The three lines under a dial -----------------------------------------------------------

    /// <summary>
    /// What a dial says underneath: the pace word, the run-out estimate, and the reset countdown.
    /// The first two hide themselves when they have nothing to say; the countdown is always there.
    /// The durations ride along so the spoken sentence says the same numbers the lines print.
    /// </summary>
    public sealed record Lines
    {
        public string? Caption { get; init; }
        public string? RunOut { get; init; }
        public required string Countdown { get; init; }
        public double? ResetSeconds { get; init; }
        public double? RunOutSeconds { get; init; }
    }

    /// <summary>
    /// The lines for one dial. A dial with no usable reset time says only that, whatever the pace
    /// thinks, matching the missing inner ring.
    ///
    /// The run-out is dropped unless it still lands before the reset: the pace is graded when a poll
    /// returns while these lines re-print every minute, so a stale Caution could otherwise project
    /// past a reset the countdown says is nearer.
    /// </summary>
    public static Lines DialLines(
        PaceStatus pace, double rawRemaining, DateTimeOffset? resetsAt, double windowSeconds, DateTimeOffset now)
    {
        if (resetsAt is not { } reset || CountdownFormat.RemainingSeconds(reset, now) is not { } resetSeconds)
        {
            return new Lines { Countdown = ResetUnavailableLine };
        }

        double? runOut = RunOutSeconds(rawRemaining, resetsAt, windowSeconds, pace, now);
        if (runOut is { } raw)
        {
            var quantised = QuantiseRunOut(raw, windowSeconds);
            runOut = quantised < resetSeconds ? quantised : null;
        }

        return new Lines
        {
            Caption = PaceCaption(pace),
            RunOut = RunOutLine(runOut),
            Countdown = "Resets in " + CountdownFormat.MinuteResolution(resetSeconds),
            ResetSeconds = resetSeconds,
            RunOutSeconds = runOut,
        };
    }

    // --- What a screen reader says ---------------------------------------------------------------

    /// <summary>
    /// A duration in words, truncated to whole units exactly like the printed countdown, so the
    /// voice never says a number the line does not show.
    /// </summary>
    public static string SpokenDuration(double seconds)
    {
        var (days, hours, minutes) = CountdownFormat.Components(seconds);
        var parts = new List<string>();
        foreach (var (value, unit) in new[] { (days, "day"), (hours, "hour"), (minutes, "minute") })
        {
            if (value > 0)
            {
                parts.Add($"{value} {unit}{(value == 1 ? string.Empty : "s")}");
            }
        }
        return parts.Count == 0 ? "less than a minute" : string.Join(" ", parts);
    }

    /// <summary>
    /// One sentence per dial (R37). Each dial is spoken once, covering how much is left, how much of
    /// the window is left, the pace, and the two lines underneath, so the text lines themselves can
    /// stay out of the accessibility tree instead of being read as four disconnected fragments.
    ///
    /// The percentages round the way the dial's own label rounds, so the voice never says a number
    /// different from the one on screen.
    /// </summary>
    public static string GaugeAccessibilityLabel(
        string name, double usage, double? timeRemaining, PaceStatus pace, Lines? lines = null)
    {
        var parts = new List<string> { $"{name} usage {(int)Math.Round(usage, MidpointRounding.ToEven)} percent" };

        if (timeRemaining is { } time)
        {
            parts.Add($"time remaining {(int)Math.Round(time, MidpointRounding.ToEven)} percent");
        }

        switch (pace)
        {
            case PaceStatus.OnTrack: parts.Add("on track"); break;
            case PaceStatus.Caution: parts.Add("caution, over pace"); break;
            case PaceStatus.Danger: parts.Add("danger, over pace"); break;
            case PaceStatus.WeeklyLimited: parts.Add("limited by weekly"); break;
        }

        if (lines is not null)
        {
            if (lines.RunOutSeconds is { } runOut)
            {
                parts.Add($"projected to run out in about {SpokenDuration(runOut)}");
            }
            parts.Add(lines.ResetSeconds is { } reset
                ? $"resets in {SpokenDuration(reset)}"
                : ResetUnavailableLine.ToLowerInvariant());
        }

        return string.Join(", ", parts);
    }
}
